File: depmesh_ai/sources/http.py
"""Tiny stdlib HTTP helper shared by all sources.

Distinguishes "the registry said 404" (authoritative: package absent)
from "the registry could not be reached" (network/policy failure) — the
vet engine treats these very differently.
"""
from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Optional

USER_AGENT = "depmesh-ai/0.1 (+https://github.com/jhberges/depmesh-ai)"
DEFAULT_TIMEOUT = 20.0


class SourceUnavailable(Exception):
    """The source could not be reached or answered abnormally."""


class NotFound(Exception):
    """The source authoritatively reported the resource as absent."""


def _open(url: str, timeout: float, accept: Optional[str]) -> bytes:
    headers = {"User-Agent": USER_AGENT}
    if accept:
        headers["Accept"] = accept
    request = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(request, timeout=timeout) as response:
            return response.read()
    except urllib.error.HTTPError as err:
        if err.code == 404:
            raise NotFound(url) from err
        raise SourceUnavailable(f"{url} -> HTTP {err.code}") from err
    except (urllib.error.URLError, OSError, TimeoutError) as err:
        raise SourceUnavailable(f"{url} -> {err}") from err


def get_json(url: str, timeout: float = DEFAULT_TIMEOUT) -> Any:
    body = _open(url, timeout, accept="application/json")
    try:
        return json.loads(body)
    except json.JSONDecodeError as err:
        raise SourceUnavailable(f"{url} -> invalid JSON") from err


def get_text(url: str, timeout: float = DEFAULT_TIMEOUT) -> str:
    return _open(url, timeout, accept=None).decode("utf-8", errors="replace")

File: internal/sources/http.go
// Package sources implements the source layer: one registry-direct module
// per ecosystem, plus optional enrichment (deps.dev). Registry modules are
// authoritative for existence; enrichment only adds signals and degrades
// gracefully when unreachable.
package sources

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"time"
)

const userAgent = "depmesh-ai/0.1 (+https://github.com/jhberges/depmesh-ai)"

// ErrNotFound means the source authoritatively reported the resource absent.
var ErrNotFound = errors.New("not found")

// UnavailableError means the source could not be reached or answered
// abnormally — the caller must treat the result as unknown, never as absent.
type UnavailableError struct {
	URL string
	Err error
}

func (e *UnavailableError) Error() string { return fmt.Sprintf("%s unavailable: %v", e.URL, e.Err) }
func (e *UnavailableError) Unwrap() error { return e.Err }

var client = &http.Client{Timeout: 20 * time.Second}

func get(url, accept string) ([]byte, error) {
	request, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, &UnavailableError{url, err}
	}
	request.Header.Set("User-Agent", userAgent)
	if accept != "" {
		request.Header.Set("Accept", accept)
	}
	response, err := client.Do(request)
	if err != nil {
		return nil, &UnavailableError{url, err}
	}
	defer response.Body.Close()
	switch {
	case response.StatusCode == http.StatusNotFound:
		return nil, fmt.Errorf("%s: %w", url, ErrNotFound)
	case response.StatusCode != http.StatusOK:
		return nil, &UnavailableError{url, fmt.Errorf("HTTP %d", response.StatusCode)}
	}
	body, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, &UnavailableError{url, err}
	}
	return body, nil
}

func getJSON(url string, into any) error {
	body, err := get(url, "application/json")
	if err != nil {
		return err
	}
	if err := json.Unmarshal(body, into); err != nil {
		return &UnavailableError{url, fmt.Errorf("invalid JSON: %w", err)}
	}
	return nil
}

func getText(url string) (string, error) {
	body, err := get(url, "")
	return string(body), err
}
